main.py

```diff
@@ -1,70 +1,93 @@
-
-
-# Output
-def save_results(col_name: str, result: str | int | float, path: str = "result.json"):
-    json_dict = {f"{col_name}": f"{result}"}
-    with open(path, "w") as fp:
-        dump(json_dict, fp, indent=4)
-
-
-print(mf_ip(df["Client IP address"]))
-print(lf_ip(df["Client IP address"]))
-print(e_ps(df["Timestamp in seconds since the epoch"]))
-#print(bytes_total(df["Response size in bytes"]))
-
-res1 = bytes_total(df["Response size in bytes"])
-
-save_results("Response size in bytes", res1)
-#save_results(bytes_total(df["Response size in bytes"]))
+#!/usr/bin/python
+
+import click
+from rich.console import Console
+
+from data_io import save_results, take_input_data
+from operations import bytes_total, e_ps, lf_ip, mf_ip
+
+console = Console()
+
+
+@click.group()
+def analyze_logs():
+    pass
+
+
+console.print(
+    """Hi :vampire:\n[b][i]This is a console log analyzer program[/i]\nIs has four operations.
+    \nIt takes an operation type argument, then location of file or a directory with logs, then directory to save result
+    \nTo use it run a command for a operation as specified below."""
+)
+
+
+@analyze_logs.command(name="mfip")
+@click.option(
+    "--mfip",
+    help=console.print(
+        "[bold magenta]Most frequent IP.[/bold magenta]",
+        "\nUse: [b]python main.py mfil[b] <input_file/dir> <output_file>",
+    ),
+    metavar="",
+)
+@click.argument("input")
+@click.argument("output")
+def mfip(mfip, input, output):
+    df = take_input_data(input)
+    mf = mf_ip(df)
+    save_results("Most frequent IP", mf, output)
+
+
+@analyze_logs.command(name="lfip")
+@click.option(
+    "--lfip",
+    help=console.print(
+        "[bold magenta]Least frequent IP.[bold magenta]",
+        "\nUse: [b]python main.py lfip[b] <input_file/dir> <output_file>",
+    ),
+    metavar="",
+)
+@click.argument("input")
+@click.argument("output")
+def lfip(lfip, input, output):
+    df = take_input_data(input)
+    lf = lf_ip(df)
+    save_results("Least frequent IP", lf, output)
+
+
+@analyze_logs.command(name="eps")
+@click.option(
+    "--eps",
+    help=console.print(
+        "[bold magenta]Events per second.[bold magenta]",
+        "\nUse: [b]python main.py eps[b] <input_file/dir> <output_file>",
+    ),
+    metavar="",
+)
+@click.argument("input")
+@click.argument("output")
+def eps(eps, input, output):
+    df = take_input_data(input)
+    eps = e_ps(df)
+    save_results("Events per second", eps, output)
+
+
+@analyze_logs.command(name="bytes")
+@click.option(
+    "--bytes",
+    help=console.print(
+        "[bold magenta]Total amount of bytes exchanged.[bold magenta]",
+        "\nUse: [b]python main.py bytes[b] <input_file/dir> <output_file>",
+    ),
+    metavar="",
+)
+@click.argument("input")
+@click.argument("output")
+def bytes(bytes, input, output):
+    df = take_input_data(input)
+    bytes = bytes_total(df)
+    save_results("Total amount of bytes exchanged", bytes, output)
+
+
+if __name__ == "__main__":
```

Comment:
wouldn't be easier to have just one command with options specified like that:
@click.command()
@click.option('--from', '-f', 'from_')
@click.option('--to', '-t')
def reserved_param_name(from_, to):
    click.echo(f"from {from_} to {to}")

Reply:
Yup, I thought about bettering main.py in that way, but didn't have enough resources to proceed with it yesterday :P 
Thank you for implementation pattern!

Reply:
After adding some features to main.py (different communicates for each commands) and first research I feel that such restructure would be too heavy in effort:result terms. Also, I don't know click that much, from implementations I've seen it is rather used linearly like in mine file, idk how to group decorators etc. Also functions would require lots of individual arguments, so making one command may not be as spectacular as expected. Also I'm kinda running out of steam, so that is to be considered as well.

Reply:
OK I went halfway through, but added cool live prepare item

def base_task(input):
    tasks = ["read file", "prepare file", "analyze and save file"]
    with console.status("[bold green]Working on tasks..."):
        while tasks:
            task = tasks.pop(0)
            sleep(1)
            console.log(f"{task}")
    df = load_input_data(input)
    return df